source/docs/software/basic-programming/functions-as-data.rst

```diff
@@ -0,0 +1,118 @@
+   Command driveHalfSpeed = runOnce(() -> { drivetrain.arcadeDrive(0.5, 0.0); }, drivetrain);
+
+Java calls ``() -> { drivetrain.arcadeDrive(0.5, 0.0); }`` a "lambda expression"; it may be less-confusingly called an "arrow function", "inline function", or "anonymous function" (because it has no name).  While this may look a bit funky, it is just another way of writing a function - the parentheses before the arrow are the function's argument list, and the code contained in the brackets is the function body.  The "lambda expression" here represents a function that calls ``drivetrain.arcadeDrive`` with a specific set of parameters - note again that this does not *call* the function, but merely defines it and passes it to the ``Command`` to be run later when the ``Command`` is scheduled.
+
+Note that our inline function still has to be a ``Runnable`` - notice that it takes no arguments and has no return statement.  If it did not match the ``Runnable`` contract, our code would fail to compile.
+
+Capturing State in Java Lambda Expressions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In the above example, our function body references an object that lives outside of the function itself (namely, the ``drivetrain`` object).  This is called a "capture" of a variable from the surrounding code (which is sometimes called the "outer scope" or "enclosing scope").  Usually the captured variables are either local variables from the enclosing method body in which the lambda expression is defined, or else fields of an enclosing class definition in which that method is defined.
+
+In Java capturing state is a fairly safe thing to do in general, with one major caveat: we can only capture state that is "effectively final".  That means it is only legal to capture a variable from the enclosing scope if that variable is never reassigned after initialization.  Note that this does not mean that the captured state cannot change: Remember that Java objects are references, so the object that the reference *points to* may change after capture - but the reference itself cannot be made to point to another object.
+
+This means we can only capture primitive types (like ``int``, ``double``, and ``boolean``) if they're constants.  If we want to capture a state variable that can change, it *must be wrapped in a mutable object*.
+
+Treating Functions as Data in C++
+---------------------------------
+
+C++ has a number of ways to treat functions as data.  For the sake of this article, we'll only talk about the parts that are relevant to using WPILibC.
+
+In WPILibC, function types are represented with the ``std::function`` class <TODO: link>.  This standard library class is templated on the function's signature - that means we have to provide it a function pointer type <TODO: link> as a template parameter to specify the signature of the function (compare this to Java above, where we have a separate interface type for each kind of signature).
```

Comment:
The `double(double)` in `std::function<double(double)>` is not a function pointer type; it's an actual function type. A function pointer type equivalent would be `double(*)(double)`.